### ISSUE TITLE
"large" gaps between the finalized block and the best-block

### ISSUE THREAD

**Issue by josepot** [smol-dot/smoldot#1754]
I'm not sure if this is a bug. However, it is a bit annoying from a UX point of view.

The issue is that there are instances in which smoldot doesn't report the `finalized` event for a ~1 minute or so. It's worth pointing out that the `bestBlock` event is being reported during that span of time. Therefore, in those instances there is a significant gap between the finalized block and the best block. I have witnessed instances with gaps of ~10 blocks.

Some logs: 
[smlogs.json](https://github.com/smol-dot/smoldot/files/14906567/smlogs.json)

For instance, in the attached logs there is a gap between: `2024-04-08T11:46:56.632Z` and `2024-04-08T11:47:52.829Z` where no `finalized` events were emitted.

Once again, I'm not sure if this is a bug or not.

Thanks,

**Comment by tomaka** [comment]
Smoldot passively receives messages from full nodes that indicate which blocks have been finalized. In the logs you give, smoldot simply didn't receive any message from the full nodes.

The situation might be helped by making smoldot actively query finality proofs.
cc https://github.com/smol-dot/smoldot/issues/1638#issuecomment-1917407636

**Comment by josepot** [comment]
> The situation might be helped by making smoldot actively query finality proofs.

That would be great.

Alternatively, we could add a [method for fetching block justifications](https://github.com/paritytech/json-rpc-interface-spec/issues/140) and -unless I'm mistaken- that would have the side-effect that smoldot would be able to determine if said block is finalized, and if it is then we would receive the `finalized` event over the chainHead subscription... is this assumption correct?

In our case we would only use said method call if we saw that we have a large amount of pinned blocks due to the fact that we haven't received the finalized event in a "long" time. For instance, if we see that there is a gap of N blocks between the best-block and the finalized one, and we also see that there are a few forks, then we could fetch the justification for the block in the `finalized + (N * (2/3))` block.

We could also have both options, of course.

**Comment by lexnv** [comment]
Once this substrate fix: https://github.com/paritytech/polkadot-sdk/pull/4135 is deployed, smoldot will receive the Grandpa Neighbor packet with every finalized block.

Then, in smoldot we could download the block justifications (and not wait for the Grandpa Commit message).
In doing so, the fetching of the justification will also help https://github.com/smol-dot/smoldot/issues/1638 and unblock subxt from updating to the latest version.

I think it would be possible to use the p2p [sync](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/network/sync/src/schema/api.v1.proto) protocol instead of adding the chainHead_justification method. Is there another way smoldot can fetch the justification of a block? (cc @tomaka)

**Comment by tomaka** [comment]
I don't understand why you both think that this is in any way related to the JSON-RPC API. This is about the syncing algorithm that smoldot uses to synchronize with the rest of the network.

**Comment by lexnv** [comment]
IIUC, the p2p [sync](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/network/sync/src/schema/api.v1.proto) protocol is the only way smoldot can fetch the justifications? 🤔

**Comment by tomaka** [comment]
Closing as duplicate of #1638
